### PR TITLE
SCIM1.1 util class

### PR DESCRIPTION
**Pull request by shavantha**
This PR is related to the SCIMProvisioningUtil
Tested environments
1.Jdk1.8.0_161
2.OS:Ubuntu 18.04

### CODE REVIEW

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+     * @param username               Authenticating username.
+     * @param password               Authenticating user password.
+     * @return HttpResponse with the result.
+     * @throws IOException If error occurs during app creation.
+     */
+
+    public static HttpResponse provisionUserSCIM11(String serverURL, JSONObject jsonObject,
+                                                   String username, String password) throws IOException {
+
+        return provisionSCIM11Entity(serverURL, "Users", jsonObject, username, password);
```
Comment (ayshsandu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: move to a constant

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.sendPostRequestWithJSON;
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.sendDeleteRequest;
+/**
+ *  Utility class for common functions for user management scenario.
+ */
+public class SCIMProvisioningUtil {
+
+    protected Log log = LogFactory.getLog(getClass());
+
+    public static final String SCIM_11_ENDPOINT = "/wso2/scim";
```
Comment (ayshsandu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: We can move this constant to a Constant.java class

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ *  Utility class for common functions for user management scenario.
+ */
+public class SCIMProvisioningUtil {
+
+
+    /**Headers
+     * @param username               Authenticating username.
+     * @param password               Authenticating user password.
+     */
+    private static Header[]  getHeader(String username, String password){
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Fix formatting issue

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.constructBasicAuthzHeader;
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.sendPostRequestWithJSON;
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.sendDeleteRequest;
+/**
+ *  Utility class for common functions for user management scenario.
+ */
+public class SCIMProvisioningUtil {
+
+
+    /**Headers
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Move `Headers` to the next line. Also let's change the method name `getCommonHeaders`

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 3 above)
Comment (omindu) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Add a new line

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * Provision a new user
+     * @param serverURL              the server url
+     * @param jsonObject             the json object of the scim payload
+     * @param username               Authenticating username.
+     * @param password               Authenticating user password.
+     * @return HttpResponse with the result.
+     * @throws IOException If error occurs during app creation.
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: during **user** creation

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+     */
+    public static HttpResponse provisionUserSCIM11(String serverURL, JSONObject jsonObject,
+                                                   String username, String password) throws IOException {
+
+        return provisionSCIM11Entity(serverURL,Constants.SCIM_ENDPOINT_USER, jsonObject, username, password);
+    }
+
+
+    /**
+     * Provision a new user
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: new `SCIM 1.1 entity`

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * Delete a  user
+     * @param serverURL              the server url.
+     * @param userId                 the ID of the user to be deleted.
+     * @param username               Authenticating username.
+     * @param password               Authenticating user password.
+     * @return HttpResponse with the result.
+     * @throws IOException If error occurs during app creation.
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Correct the description

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+     */
+    public static HttpResponse deleteUser(String serverURL,String userId,
+                                          String username, String password) throws IOException {
+
+        return deleteUserEntity(serverURL,userId,username, password);
+    }
+
+    /**
+     * Delete a  user
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Fix teh description

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                                                   String username, String password) throws IOException {
+
+        HttpClient client = HttpClients.createDefault();
+        String scimEndpoint = serverURL + Constants.SCIM_11_ENDPOINT + "/" + scim11EndpointType;
+
+        return sendPostRequestWithJSON(client, scimEndpoint, jsonObject,getHeader(username, password));
+    }
+
+    /**
+     * Delete a  user
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Additional space before `user`

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+     * @param serverURL              the server url
+     * @param userID                 the ID of the user to be deleted
+     * @param username               Authenticating username.
+     * @param password               Authenticating user password.
+     * @return HttpResponse with the result.
+     * @throws IOException If error occurs during app creation.
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Fix description

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+     */
+    public static HttpResponse deleteUserEntity(String serverURL, String userID,
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Change this to `deleteSCIM11Entity`. Add an argument to get the endpoint type (Users/Groups) similar to `provisionSCIM11Entity`

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+     * @param userId                 the ID of the user to be deleted.
+     * @param username               Authenticating username.
+     * @param password               Authenticating user password.
+     * @return HttpResponse with the result.
+     * @throws IOException If error occurs during app creation.
+     */
+    public static HttpResponse deleteUser(String serverURL,String userId,
+                                          String username, String password) throws IOException {
+
+        return deleteUserEntity(serverURL,userId,username, password);
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Fix formatting.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+import org.apache.http.client.HttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
+import org.json.simple.JSONObject;
+import java.io.IOException;
+
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.constructBasicAuthzHeader;
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.sendPostRequestWithJSON;
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.sendDeleteRequest;
+/**
```
Comment (omindu) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Add a new line.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/Constants.java [hunk 14]
```diff
@@ -52,4 +52,6 @@ private Constants() {
     public static final String OAUTH_AUTHORIZE_URI_CONTEXT = "/oauth2/authorize";
     public static final String OAUTH_TOKEN_URI_CONTEXT = "/oauth2/token";
     public static final String COMMONAUTH_URI_CONTEXT = "/commonauth";
+    public static final String SCIM_11_ENDPOINT = "/wso2/scim";
+    public static final String SCIM_ENDPOINT_USER="Users";
```
[the remark this answers is not in the document]
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/Constants.java]: done, included a space as SCIM_ENDPOINT_USER = "Users"

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 13 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: done, added a new line between lines 30 and 31

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+ *  Utility class for common functions for user management scenario.
+ */
+public class SCIMProvisioningUtil {
+
+
```
[the remark this answers is not in the document]
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: done, removed the extra lines as mentioned

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 4 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: done, updated the relevant areas

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 9 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: done, removed the extra space

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 5 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: thanks for pointing out fixed the description

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 6 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: thanks, fixed the description

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 7 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: thanks, fixed all lines which read "app creation"

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/Constants.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/Constants.java]: thanks, fixed the spacing issue on line 56

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 10 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Done, improved the description and the exception related description by removing the app creation and replaced with the accurate wording

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 3 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: done, added a new line between comments and private static Header

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 12 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: done, fixed the formatting on the methods parameter spacing

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 11 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: in both provisionSCIM11Entity and deleteSCIM11Entity a change was made where the endpoint type is not passed as a parameter but is taken from Constants class and passed to the URL when constructing the URL.

provisionSCIM11Entity
String scimEndpoint = serverURL + Constants.SCIM_11_ENDPOINT + "/" + Constants.SCIM_ENDPOINT_USER;

deleteSCIM11Entity
String scimEndpoint = serverURL + Constants.SCIM_11_ENDPOINT + "/" + Constants.SCIM_ENDPOINT_USER+ "/" + userID;

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+     *  Utility class for common functions for user management scenario.
+     */
+public class SCIMProvisioningUtil {
+
+    /**
+     * Headers
+     * @param username    Authenticating username.
+     * @param password    Authenticating user password.
+     */
+    private static Header[]  getCommonHeaders(String username, String password){
```
[the remark this answers is not in the document]
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: done, added the changes

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+     * Provision a new user
+     * @param serverURL              the server url
+     * @param jsonObject             the json object of the scim payload
+     * @param username               Authenticating username.
+     * @param password               Authenticating user password.
+     * @return HttpResponse with the result.
+     * @throws IOException If error occurs during user creation.
+     */
+    public static HttpResponse provisionUserSCIM11(String serverURL, JSONObject jsonObject,
+                                                   String username, String password) throws IOException {
```
[the remark this answers is not in the document]
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: done,added a new line between the comments and the public static statement

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java (hunk 11 above)
Comment (shavantha) [reply product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/util/SCIMProvisioningUtil.java]: Done, introduced this change now the test class needs to pass the serverURL and the SCIM endpoint type to the util class. From the scimEndPoint parameter we can pass ,wso2scim for scim 1.1 or scim2 for SCIM2.0 to construct the URL.

**Pull request merged**